implement streamDownload() returning a stream

danielecr opened the pull request.

danielecr commented:
The usage of streamDownload() is not as simple as downloadReport(), the createReport and getReportDocument steps must be done before doing:

const docStream = await this.sellingPartner.downloadStream(report_document);

this returns a readableStream, but actually it is the response object from https.request, that in turns implements the stream API: from the programmer point of view this is the same.

What you can do with a ReadableStream, can be done with docStream, ex:

> docStream.pipe(fileWriter);

where fileWriter is a stream writing in a file.

danielecr commented:
@amz-tools I need review for this patch it solves #56 in a more elegant way

jesusvalle commented:
This feature would be great. Up!

amz-tools commented:
@danielecr @jesusvalle Will try to review as soon as possible so that we can integrate it in the next version.

amz-tools commented:
@danielecr I finally had the chance to take a look at the implementation. I made some changes, see here: https://github.com/amz-tools/amazon-sp-api/tree/pull-requests/306

I realized that the downloadStream works fine, but only if the report is not gzipped, which is true for `GET_FLAT_FILE_OPEN_LISTINGS_DATA` report. However trying the same with i.e. `GET_V2_SELLER_PERFORMANCE_REPORT` report did not work. I fixed this by adding a check on compressionAlgorithm presence in downloadStream and if its there pipe it through zlib.createGunzip():

`return details.compressionAlgorithm ? res.pipe(zlib.createGunzip()) : res;`

I also noticed some issues with the test you added, please see the TODO comments.

One more thing I added is a `downloadReportStream()` function. I think almost everyone currently uses the `downloadReport()` (which I think wasn't available when you first starting with implementing the streaming functionality) instead of `download()` because its much more convenient as it internally wraps all the reporting functions (`createReport`, `getReport`, `getReportDocument`). So I thought this might be absolutely helpful to implement it for streams as well.

Also maybe we should try to include the timeout functionality for the downloadStream as well?

And I haven't tested/looked at the uploadStream functionality yet. What I was wondering: Do we really need it? Whats might be a usecase for it? 

Could you take a look at the changes I implemented and also give some feedback on the open topics? Thanks!

jesusvalle commented:
Hi @danielecr ! Could you please provide an example of how to use the uploadStream method?

I'm trying to upload a JSON_LISTINGS_FEED using it, but I get an error saying that it can't read statusCode of undefined.

Thanks a lot!

danielecr commented:
> @danielecr I finally had the chance to take a look at the implementation. I made some changes, see here: https://github.com/amz-tools/amazon-sp-api/tree/pull-requests/306
> 
> I realized that the downloadStream works fine, but only if the report is not gzipped, which is true for `GET_FLAT_FILE_OPEN_LISTINGS_DATA` report. However trying the same with i.e. `GET_V2_SELLER_PERFORMANCE_REPORT` report did not work. I fixed this by adding a check on compressionAlgorithm presence in downloadStream and if its there pipe it through zlib.createGunzip():
> 
> `return details.compressionAlgorithm ? res.pipe(zlib.createGunzip()) : res;`

This is a questionable choice, a possible use case can be:

```
const docStream = await this.sellingPartner.downloadStream(report_document);
const target_filename = 'basefile.csv' + details.compressionAlgorithm ? '.gz' : '';
```

and then `.pipe()` into the outputStream with right filename extension.

Some csv reading tools deal very well with compressed data, and also perform better.


> 
> I also noticed some issues with the test you added, please see the TODO comments.
> 
> One more thing I added is a `downloadReportStream()` function. I think almost everyone currently uses the `downloadReport()` (which I think wasn't available when you first started with implementing the streaming functionality) instead of `download()` because its much more convenient as it internally wraps all the reporting functions (`createReport`, `getReport`, `getReportDocument`). So I thought this might be absolutely helpful to implement it for streams as well.
> 
> Also maybe we should try to include the timeout functionality for the downloadStream as well?

Well, I also thought about it. The biggest difference here is on the type of timeout, and the way one would deal with it.

Also I found this:

https://stackoverflow.com/questions/42301512/timeout-handling-with-node-js-stream-piping

it is a old question, but I trust it.

My idea was that basically this call expose the lower level https.request() stream and event handler, so any addition would requires additional cognitive load, or define a new kind of standard that is not really required.

On other hand, a new dependency can be added, I mean some well known stream utility from:
https://github.com/thejmazz/awesome-nodejs-streams

but at this point it is out-of-scope, because it can be installed together with the amazon-sp-api npm package, and used at your will

> 
> And I haven't tested/looked at the uploadStream functionality yet. What I was wondering: Do we really need it? Whats might be a usecase for it?

upload pdf file for IDU can be a use case.

In any case stream interface is useful when the library is used on a "shared" environment,  stream also does not load all data in memory, and this has a good impact on security

> 
> Could you take a look at the changes I implemented and also give some feedback on the open topics? Thanks!

I will do it.

jesusvalle commented:
> > Hi @danielecr ! Could you please provide an example of how to use the uploadStream method?
> > I'm trying to upload a JSON_LISTINGS_FEED using it, but I get an error saying that it can't read statusCode of undefined.
> > Thanks a lot!
> 
> Thank you for testing, I will try it and fix

Thanks. That would be great because, in this type of feed, a very large volume of data can accumulate, and the ability to upload it using the uploadStream method would solve many problems.
